designs/2021-schema-object-rules/README.md

```diff
@@ -0,0 +1,284 @@
+- Repo: eslint/eslint and @eslint/eslintrc
+- Start Date: 2021-12-12
+- RFC PR: <https://github.com/eslint/rfcs/pull/85>
+- Authors: [bmish](https://github.com/bmish)
+
+# Require schemas and object-style rules
+
+## Summary
+
+<!-- One-paragraph explanation of the feature. -->
+
+We propose two changes to rules:
+
+- Requiring rules with options to specify schemas (removing support for rules with options that are missing schemas)
+- Requiring rule implementations to use the object-style format (removing support for the deprecated function-style format)
+
+Note that while these two proposals are included in the same RFC so that they can be considered holistically, they do not have to be tied together, and can be split into separate RFCs if desired.
+
+## Motivation
+
+<!-- Why are we doing this? What use cases does it support? What is the expected
+outcome? -->
+
+### Motivation for requiring schemas
+
+Today, rule schemas ([meta.schema](https://eslint.org/docs/developer-guide/working-with-rules#rule-basics)) are optional. This has the following effects:
+
+- If a rule does not specify a schema, nothing is enforced.
+- Rules with options may not have a schema specified, despite it being best practice to specify a schema for rule options. This increases the chances of developers passing invalid rule options and encountering unexpected behavior / bugs in rules.
+- If a rule author wants to enforce that no options are passed to their rule, they have to manually specify `schema: []`, otherwise it's possible that developers could accidentally provide useless rule options to their rule without knowing it.
+
+So we will change the default rule schema to `schema: []` to enforce that no rule options are passed to rules that do not specify a schema.
+
+This has the desirable consequence of requiring rules with options to begin specifying their schema if they did not already.
+
+Making ESLint's default behavior more strict like this goes along with many recent changes to tighten validation (i.e. more [RuleTester class validation in ESLint 7](https://eslint.org/docs/user-guide/migrating-to-7.0.0#rule-tester-strict), [increased rule configuration validation in ESLint 6](https://eslint.org/docs/user-guide/migrating-to-6.0.0#rule-config-validating), etc) to improve rule quality and usability and reduce the chance of mistakes.
+
+And in addition to the obvious benefits of increased rule schema usage such as reduced chance of silent mistakes/bugs, being able to rely on having rule schemas available could unlock new features/improvements:
+
+- Improved/auto-generated documentation regarding rule options
+- TypeScript-style auto-complete/IntelliSense when configuring a rule
+
+There is an existing, third-party lint rule [eslint-plugin/require-meta-schema](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/require-meta-schema.md) to enforce that rules have schemas.
```

Comment:
I'm 👍to change the default to ` []`, but do we want to add a new way to allow rule authors explicitly skipping the validating - e.g. `schema: false`?
I'm asking as it could happen in some rare cases, e.g. writing an internal rule.

Reply:
> I'm to change the default to ` []`, but do we want to add a new way to allow rule authors explicitly skipping the validating - e.g. `schema: false`?

A workaround can be `schema: {}` or `schema: { type: "array" }`

Reply:
I could see both sides, but my current preference is to *not* provide an explicit opt-out.

Why we shouldn't give users an explicit schema opt-out:
- An easy, well-known option to opt-out is just welcoming laziness, and many plugins authors may default to always using it, despite the benefits and ease of writing a real schema, defeating many of the benefits described in this RFC
- An explicit opt-out option is essentially encouraging/recommending opting-out as an acceptable approach
- As @mdjermanovic suggested, users can use the the workaround of providing an incomplete schema like`schema: {}` or `schema: { type: "array" }` if they really want to opt-out
- In my opinion, schemas should be low-effort and valuable for all rules, even internal rules

Why we should give users an explicit schema opt-out like `schema: false`:
- If users are going to abuse obscure workarounds like `schema: {}` or `schema: { type: "array" }` anyway, then it would be clearer for everyone if they could explicitly specify that they don't want a schema

Reply:
I added a paragraph about this workaround in the `Drawbacks of requiring schemas` section including touching on why I am against an explicit opt-out.

Reply:
Both arguments for and arguments against `schema: false` are good.

> Why we should give users an explicit schema opt-out like `schema: false`:
> 
> * If users are going to abuse obscure workarounds like `schema: {}` or `schema: { type: "array" }` anyway, then it would be clearer for everyone if they could explicitly specify that they don't want a schema

I suggested the workarounds, but this is a very good argument and I'm now for an explicit opt-out like `schema: false`.

As described in the Motivation section of this RFC, one of the benefits of rule schemas (and thus forcing rules to define schemas) is automated analysis of rule options, like auto-generated docs, auto-complete, maybe even detecting breaking changes as mentioned in https://github.com/eslint/eslint/issues/15428, etc., For that purpose, when some rule doesn't provide a schema for whatever reason, it seems better to see that right away from `schema: false` than to analyze a meaningless workaround schema.

Also, some schema validations we might want to add to RuleTester in the future could report these workaround schemas as potentially incorrect, since they're essentially a no-op.

Reply:
Also in favor of an explicit `schema: false` opt-out. I can imagine some companies' internal rules that aren't accessible to the (excellent, incredibly helpful) ecosystem analyzer don't have schemas because they're only configured once by the person who wrote the rule. Having `schema: false` means `eslint-plugin-eslint-plugin` could trivially add a warning-level rule to encourage finishing the schema rather than having to analyze no-op schemas.

Reply:
Sounds good, I switched to use `schema: false`.

Reply:
One follow-up to this, I think we should disallow `schema: {}`. That is the simplest no-op schema and can be easily confused with `schema: []` which actually does enforce that there are no options. Anyone using `schema: {}` is likely making a mistake, so we can throw an error if we detect that:

```
`schema: {}` is a no-op. For rules with options, please fill in a complete schema. For rules without options, please omit `schema` or use `schema: []`.
```

We could also disallow other obvious ones like `schema: { type: "array" }` but I'm not too worried about catching all possible no-op schemas right now.

Reply:
`schema: { additionalProperties: false }` is also a no-op.

Reply:
ah, maybe `{ maxProperties: 0 }` or something, but clearly `[]` is simpler

Reply:
I updated the `Opt-out` section to mention that we will disallow a few common no-op schemas.